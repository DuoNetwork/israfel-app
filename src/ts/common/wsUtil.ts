import WebSocket from 'isomorphic-ws';
import * as CST from '../../../../israfel-relayer/src/common/constants';
import {
	IWsOrderResponse,
	IWsUserOrderResponse
} from '../../../../israfel-relayer/src/common/types';
import Web3Util from '../../../../israfel-relayer/src/utils/Web3Util';
import { IWsAddOrderRequest } from '../common/types';
import util from './util';

class WsUtil {
	public ws: WebSocket | null = null;
	private web3Util: Web3Util | null = null;
	private handleOrderBooksUpdate: ((orderBooks: IWsUserOrderResponse) => any) | null = null;
	public init(host: string) {
		this.web3Util = new Web3Util(window, !__DEV__);
		this.ws = new WebSocket(host);
		this.ws.onopen = function open() {
			console.log('Connected');
		};
		this.ws.onmessage = (m: any) => this.handleMessage(m.data.toString());
	}

	public handleMessage(message: string) {
		const wsMsg: IWsOrderResponse = JSON.parse(message);
		const method = wsMsg.method;
		switch (method) {
			case CST.DB_ADD:
				if (this.handleOrderBooksUpdate) {
					const obRes = wsMsg as IWsUserOrderResponse;
					// const src = others[0];
					if (obRes && obRes.userOrder) {
						// obRes.delay = util.getUTCNowTimestamp() - obRes.timestamp;
						console.log('obRes');
						this.handleOrderBooksUpdate(obRes);
					}
				}
				break;
			default:
				break;
		}
	}

	public subscribe(name: string, marketId: string) {
		if (this.ws && name && marketId) {
			const req = {
				type: 'subscribe',
				channel: {
					name: name,
					marketId: marketId
				},
				requestId: Date.now().toString()
			};
			this.ws.send(JSON.stringify(req));
		}
	}

	public async addOrder(zrxAmt: number, ethAmt: number, isBid: boolean) {
		if (!this.web3Util) throw new Error('error');
		const zrxTokenAddress = this.web3Util.getTokenAddressFromName(CST.TOKEN_ZRX);
		const etherTokenAddress = this.web3Util.getTokenAddressFromName(CST.TOKEN_WETH);
		if (etherTokenAddress === undefined) throw console.error('undefined etherTokenAddress');

		const accounts = await this.web3Util.web3Wrapper.getAvailableAddressesAsync();
		const rawOrder = await this.web3Util.createRawOrder(
			accounts[0],
			__DEV__ ? CST.RELAYER_ADDR_KOVAN : CST.RELAYER_ADDR_MAIN,
			isBid ? zrxTokenAddress : etherTokenAddress,
			isBid ? etherTokenAddress : zrxTokenAddress,
			isBid ? zrxAmt : ethAmt,
			isBid ? ethAmt : zrxAmt,
			Math.ceil(util.getUTCNowTimestamp() / 1000) + 3600
		);
		const pair = 'ZRX-WETH';
		const msg: IWsAddOrderRequest = {
			method: CST.DB_ADD,
			channel: CST.DB_ORDERS,
			pair: pair,
			orderHash: rawOrder.orderHash,
			order: rawOrder.signedOrder
		};
		this.ws.send(JSON.stringify(msg));
	}

	public onOrderBooks(handleOrderBooksUpdate: (orderBooks: IWsUserOrderResponse) => any) {
		this.handleOrderBooksUpdate = handleOrderBooksUpdate;
	}
}

const wsUtil = new WsUtil();
export default wsUtil;
