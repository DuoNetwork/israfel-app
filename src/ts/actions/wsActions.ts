import * as CST from 'ts/common/constants';
import {
	IAcceptedPrice,
	IOrderBookSnapshot,
	IOrderBookSnapshotUpdate,
	IPrice,
	IStatus,
	IToken,
	IUserOrder,
	VoidThunkAction
} from 'ts/common/types';
import wsUtil from 'ts/common/wsUtil';

export function connectionUpdate(connected: boolean) {
	return {
		type: CST.AC_CONNECTION,
		value: connected
	};
}

export function infoUpdate(
	tokens: IToken[],
	status: IStatus[],
	acceptedPrices: { [custodian: string]: IAcceptedPrice[] },
	exchangePrices: { [source: string]: IPrice[] }
) {
	return {
		type: CST.AC_INFO,
		tokens: tokens,
		status: status,
		acceptedPrices: acceptedPrices,
		exchangePrices: exchangePrices
	};
}

export function orderUpdate(order: IUserOrder) {
	return {
		type: CST.AC_ORDER,
		value: order
	};
}

export function orderHistoryUpdate(orderHistory: IUserOrder[]) {
	return {
		type: CST.AC_ORDER_HISTORY,
		value: orderHistory
	};
}

export function orderBookUpdate(obUpdate: IOrderBookSnapshotUpdate) {
	return {
		type: CST.AC_OB_UPDATE,
		value: obUpdate
	};
}

export function orderSubscriptionUpdate(account: string) {
	return {
		type: CST.AC_OB_SUB,
		account: account
	};
}

export function orderBookSubscriptionUpdate(pair: string) {
	return {
		type: CST.AC_OB_SUB,
		pair: pair
	};
}

export function orderBookSnapshotUpdate(orderBook: IOrderBookSnapshot) {
	return {
		type: CST.AC_OB_SNAPSHOT,
		value: orderBook
	};
}

export function subscribeOrderBook(pair: string): VoidThunkAction {
	return dispatch => {
		dispatch(orderBookSubscriptionUpdate(''));
		dispatch(orderBookSubscriptionUpdate(pair));
		wsUtil.subscribeOrderBook(pair);
	};
}
