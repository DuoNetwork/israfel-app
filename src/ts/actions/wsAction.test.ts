import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
// import * as CST from 'ts/common/constants';
import wsUtil from 'ts/common/wsUtil';
import * as wsActions from './wsActions';

const mockStore = configureMockStore([thunk]);

describe('actions', () => {
	test('connectionUpdate', () => {
		expect(wsActions.connectionUpdate(true)).toMatchSnapshot();
	});

	test('infoUpdate', () => {
		expect(
			wsActions.infoUpdate(
				['token1'] as any,
				['status1'] as any,
				{
					custodian: ['acceptPrices'] as any
				},
				{
					source: ['exchangePrices'] as any
				}
			)
		).toMatchSnapshot();
	});

	// test('orderUpdate', () => {
	// 	expect(wsActions.orderUpdate({ test: 'test' } as any)).toMatchSnapshot();
	// });

	// test('orderHistoryUpdate', () => {
	// 	expect(dexActions.orderHistoryUpdate([{ test: 'test' }] as any)).toMatchSnapshot();
	// });

	test('orderBookSnapshotUpdate', () => {
		expect(wsActions.orderBookSnapshotUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('orderBookUpdate', () => {
		expect(wsActions.orderBookUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('orderBookSubscriptionUpdate', () => {
		expect(wsActions.orderBookSubscriptionUpdate('pair')).toMatchSnapshot();
	});

	test('subscribe', () => {
		window.setInterval = jest.fn(() => 123);
		const store: any = mockStore({});
		wsUtil.subscribeOrderBook = jest.fn();
		store.dispatch(wsActions.subscribe('pair'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect((wsUtil.subscribeOrderBook as jest.Mock).mock.calls).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
