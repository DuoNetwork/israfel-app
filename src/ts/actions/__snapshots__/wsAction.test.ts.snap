// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions connectionUpdate 1`] = `
Object {
  "type": "connection",
  "value": true,
}
`;

exports[`actions infoUpdate 1`] = `
Object {
  "acceptedPrices": Object {
    "custodian": Array [
      "acceptPrices",
    ],
  },
  "exchangePrices": Object {
    "source": Array [
      "exchangePrices",
    ],
  },
  "status": Array [
    "status1",
  ],
  "tokens": Array [
    "token1",
  ],
  "type": "info",
}
`;

exports[`actions orderBookSnapshotUpdate 1`] = `
Object {
  "type": "orderBookSnapshot",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions orderBookSubscriptionUpdate 1`] = `
Object {
  "pair": "pair",
  "type": "orderBookSubscription",
}
`;

exports[`actions orderBookUpdate 1`] = `
Object {
  "type": "orderBookUpdate",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions orderHistoryUpdate 1`] = `
Object {
  "type": "orderHistory",
  "value": Array [
    Object {
      "test": "test",
    },
  ],
}
`;

exports[`actions orderSubscriptionUpdate 1`] = `
Object {
  "account": "account",
  "type": "orderBookSubscription",
}
`;

exports[`actions orderUpdate 1`] = `
Object {
  "type": "order",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions subscribeOrderBook 1`] = `
Array [
  Object {
    "pair": "",
    "type": "orderBookSubscription",
  },
  Object {
    "pair": "pair",
    "type": "orderBookSubscription",
  },
]
`;

exports[`actions subscribeOrderBook 2`] = `
Array [
  Array [
    "pair",
  ],
]
`;
